### DISCUSSION TITLE
handbook/0.1/atas/20230710_publicacao_dados_comissao_etica/

### DISCUSSION THREAD

**Discussion by giscus** [transparencia-mg/handbook#74] [Announcements]
# handbook/0.1/atas/20230710_publicacao_dados_comissao_etica/

https://transparencia-mg.github.io/handbook/0.1/atas/20230710_publicacao_dados_comissao_etica/

**Comment by AntonioMarcel** [comment]
Estive trabalhando essa semana numa documentação mais detalhada dos passos de inicialização do projeto via terminal e achei interessante compartilhar. Segue uma série de posts com o material que redigi. Pretendo melhorá-lo no futuro implementando linguagem de markdown.

**Comment by AntonioMarcel** [comment]
#Instalar terminal git
https://git-scm.com/downloads

#criar organização conset no github
#criar repositório de treinamento no github
#boas práticas: criar organizações, repositórios e projetos utilizando - (dash) ao invés de _ (underline). Evita possíveis erros.

python --version #versão do python instalado

mkdir code\conset-cge-mg #cria pasta da organização
mkdir treinamento-dados-abertos #cria pasta do treinamento

cd code\conset-cge-mg\treinamento-dados-abertos #acessar pasta do treinamento

touch README.md #cria o arquivo README
touch .gitignore #cria o arquivo gitignore (contém os itens que serão ignorados pelo git e não serão comitados)
touch .env #cria o arquivo .env (conterá as variáveis de login no ambiente de dados abertos)

#(New-item: comando equivalente ao touch no Windows)
#(Remove-Item: comando equivalente ao rm no Windows)

git --version #versão do git instalado

#configurações iniciais
git init #inicializa o git (inicializa o repositório vazio .git)

#cria o ambiente virtual python para o repositório
#a respeito do ambiente virtual: para cada repositório/projeto, cria-se um novo ambiente virtual. Isso facilita o versionamento dos pacotes e evita conflitos
conda deactivate #se no PowerShell: desativa o ambiente conda
python -m venv venv #cria o ambiente virtual
. venv/scripts/activate #ativa o ambiente virtual

#logar no git (necessário para fazer os commits)
git config --global user.email "xxxxx@email.com" #email do github
git config --global user.name "xxxxxx" #nome de usuário do github

#mostra arquivos adicionados/alterados e a serem comitados (lembrando que o .gitignore faz os arquivos incluídos nele serem ignorados na hora do commit))
git status

#adiciona os arquivos para commit (se quiser adicionar arquivo por arquivo e fazer commit por commit, substituir . pelo nome do arquivo)
#também é possível fazer o git add para múltiplos arquivos, especificando-os (especifica o nome de cada um no comando)
#notação .\ significa "procure no diretório atual" (foi utilizada na demonstração pelo Gabriel)
git add .

#faz o commit. Boa prática: sempre comentar os commits e iniciar os comentários com verbo no presente do indicativo
git commit -m "faz commit inicial"

git log #mostra lista de commits

git branch #lista todos os branches
git branch -M main #muda o nome do branch atual (master) para main (convenção)

#cria o alias origin para o repositório para poder subir os arquivos para o github
git remote add origin https://github.com/conset-cge-mg/treinamento-dados-abertos.git

git remote -v #lista as conexões remotas criadas e a URL de cada uma

git push origin main #joga os arquivos da branch main para o github referentes ao último commit realizado

#cria o arquivo requirements que conterá os pacotes a ser instalados (adicionar pacotes a serem instalados)
touch requirements.txt
pip list #mostra os pacotes instalados
pip install -r requirements.txt #instala os pacotes listados em requirements

git add requirements.txt #adiciona o arquivo requirements.txt para commit
git commit -m "requirements.txt comitado" #comita o arquivo requirements.txt

git push origin main #joga os arquivos da branch main para o github referentes ao último commit realizado

#cria o arquivo data.csv (cria as colunas e lança alguns registros para servir de base na hora de exportar)
#usa o frictionless (instalado a partir do dpckan) para descrever os dados do pacote e exportar para o arquivo datapackage.yaml
frictionless describe data.csv --type package > datapackage.yaml

#valida o datapackage .yaml (importante sempre validar quando ocorrer alguma alteração para garantir que não há nada de errado)
#gerando erro no PowerShell: erro de codificação (utf-8). Salvar manualmente para evitar esse erro
frictionless validate datapackage.yaml

#uma vez que o arquivo .yaml é validado, exporta para .json
frictionless describe datapackage.yaml --type package --json > datapackage.json

#valida o datapackage .json
#gerando erro no PowerShell: erro de codificação (utf-8). Salvar manualmente para evitar esse erro
frictionless validate datapackage.json

#faz os commits dos arquivos criados/alterados
git add data.csv
git commit -m "cria o dataset"
git add datapackage.yaml
git commit -m "valida o arquivo .yaml"
git add datapackage.json
git commit -m "valida o arquivo .json inicial"

git push origin main #joga os arquivos para o github referentes aos commits realizados

#subir os arquivos do projeto para o ambiente de homologação
#fazer login
https://homologa.cge.mg.gov.br

#colocar as chaves de acesso no arquivo .env
#criar os secrets no settings do repositório do github???
#para evitar erro, não esquece de incluir a propriedade title no datapackage (debugado por meio do package ipdb)

dpckan --datastore dataset create #cria a estrutura de dados no homologa

#criação e utilização dos tasks (biblioteca taskipy)
touch pyproject.toml #cria o arquivo com os alias e tasks
task --list #lista as tasks disponíveis

git add . #alterações em requirements e pyproject.toml
git commit -m "adiciona taskipy e cria dois tasks"
git push origin main #joga os arquivos para o github referentes aos commits realizados

#faz alterações no arquivo .yaml
task datapackage-json #exporta o arquivo .json após alterações
task validate #valida o novo arquivo .json criado
task dataset-update #atualiza a instância do CKAN no homologa

git add .
git commit -m "altera datapackage e atualiza ckan no homologa"
git push origin main #joga os arquivos para o github referentes aos commits realizados

**Comment by AntonioMarcel** [comment]
requirements.txt
dpckan: para subir os arquivos pro homologa
ipdb: para debugar o código
taskipy: criar alias pras tarefas no git

**Comment by AntonioMarcel** [comment]
pyproject.toml
[tool.taskipy.tasks]
datapackage-json = { cmd = "frictionless describe datapackage.yaml --type package --json > datapackage.json", help = "Converte datapackage.yaml em datapackage.json." }
dataset-update= { cmd = "dpckan --datastore dataset update", help = "Atualiza CKAN." }
validate= { cmd = "frictionless validate datapackage.json", help = "Valida datapackage.json." }

**Comment by AntonioMarcel** [comment]
Documentação mais detalhada de alguns passos e conceitos:

Commit dos arquivos datapackage:
comitados os arquivos datapackage (commits: validados os arquivos...)
erros para criar no homologa (faltou propriedades nos datapackages (name e owner_org)) -> alterações nos arquivos datapackage
novo commit dos arquivos datapackage com as alterações feitas (commit: gera datapackage)
mesmo assim ainda não estava subindo os arquivos pelo dpckan
debugado o código e identificado que faltou a propriedade title -> altera-se os arquivos
aceitou a criação do dataset pelo dpckan
feito novo commit (commit: atualiza datapackage)

Pasta temp:
Quando o dpckan roda ele cria uma pasta temp com arquivo do datapackage em formato .json no início e depois deleta esse arquivo no final.
Só que ao tentar debugar o código para entender porque não conseguimos subir o dataset pro homologa pelo dpckan, usamos o pacote ipdb que faz com que o código seja rodado até certo ponto que definimos. Aí o arquivo temp acabou sendo criado e permanecendo, porque o código do dpckan não rodou até o final.
Aí mesmo depois de ter dado certo de subir os dados pro homologa pelo dpckan, o temp continuou lá. Aí fizemos um commit para isso e colocamos a pasta temp no .gitignore.

Sobre o ambiente virtual:

What does VENV do?
The virtualenv creates a folder which contains all the necessary executables to use the packages that a Python project would need. So, each project can have its own dependencies, regardless of what dependencies every other project has.
https://www.bogotobogo.com/python/python_virtualenv_virtualenvwrapper.php#:~:text=The%20virtualenv%20creates%20a%20folder,dependencies%20every%20other%20project%20has.

Criação/atualização de dados no DPCKAN:
Ao usar os comandos create e update o DPCKAN pega os arquivos .json criados a partir do dataset e os joga para a instância do homologa.

Possíveis melhorias no DPCKAN:
mostrar mensagem de erro para inserir propriedade title no datapackage (depois abrir issue no pacote do dpckan)

**Comment by AntonioMarcel** [comment]
Dúvidas:

1. Colocar as chaves de acesso no arquivo .env (pedir pra mostrar de novo como puxar essas chaves de acesso???)
2. Commit pasta temp no gitignore???
3. Precisa jogar o código do taskipy no terminal ou é só criar o arquivo toml no projeto???
4. criar os secrets no settings do repositório do github???

Respostas:

1. Detalhado no dia que fizemos fork e clone do projeto new-dataset-template. Também foi gravado um vídeo de referência.
https://youtu.be/VbsRpYB1G-U
2. Detalhado no tópico acima
3. Não precisa. É só fazer as tasks e jogar no arquivo toml.
